src/software/sensor_fusion/sensor_fusion.cpp

```diff
@@ -0,0 +1,86 @@
+
+SensorFusion::SensorFusion()
+    : field_state(0, 0, 0, 0, 0, 0, 0, Timestamp::fromSeconds(0)),
+      ball_state(Point(), Vector(), Timestamp::fromSeconds(0)),
+      friendly_team_state(Duration::fromMilliseconds(
+          Util::Constants::ROBOT_DEBOUNCE_DURATION_MILLISECONDS)),
+      enemy_team_state(Duration::fromMilliseconds(
+          Util::Constants::ROBOT_DEBOUNCE_DURATION_MILLISECONDS)),
+      ball_filter(BallFilter::DEFAULT_MIN_BUFFER_SIZE,
+                  BallFilter::DEFAULT_MAX_BUFFER_SIZE),
+      friendly_team_filter(),
+      enemy_team_filter()
+{
+}
+
+void SensorFusion::onValueReceived(RefboxData refbox_data)
+{
+    updateWorld(refbox_data);
+    Subject<World>::sendValueToObservers(world);
+}
+
+void SensorFusion::onValueReceived(RobotStatus robot_status)
+{
+    updateWorld(robot_status);
+    Subject<World>::sendValueToObservers(world);
+}
+
+void SensorFusion::onValueReceived(VisionDetection vision_detection)
+{
+    updateWorld(vision_detection);
+    Subject<World>::sendValueToObservers(world);
+}
+
+void SensorFusion::updateWorld(RefboxData refbox_data)
+{
+    world.updateRefboxData(refbox_data);
+}
+
+void SensorFusion::updateWorld(RobotStatus robot_status) {}
+
+void SensorFusion::updateWorld(VisionDetection vision_detection)
+{
+    // Parse Detections
+    std::vector<BallDetection> ball_detections = vision_detection.getBallDetections();
+    std::vector<RobotDetection> friendly_robot_detections =
+        vision_detection.getFriendlyTeamDetections();
+    std::vector<RobotDetection> enemy_robot_detections =
+        vision_detection.getEnemyTeamDetections();
+    std::optional<Field> field_detection = vision_detection.getFieldDetection();
+
+    // Apply filters
+    std::optional<Ball> new_ball =
+        ball_filter.getFilteredData(ball_detections, field_state);
+    Team new_friendly_team = friendly_team_filter.getFilteredData(
+        friendly_team_state, friendly_robot_detections);
+    Team new_enemy_team =
+        enemy_team_filter.getFilteredData(enemy_team_state, enemy_robot_detections);
+
+    // Set team goalies
+    int friendly_goalie_id = Util::DynamicParameters->getAIControlConfig()
+                                 ->getRefboxConfig()
+                                 ->FriendlyGoalieId()
+                                 ->value();
+    new_friendly_team.assignGoalie(friendly_goalie_id);
+    int enemy_goalie_id = Util::DynamicParameters->getAIControlConfig()
+                              ->getRefboxConfig()
+                              ->EnemyGoalieId()
+                              ->value();
+    new_enemy_team.assignGoalie(enemy_goalie_id);
+
+    // Update World
+    if (field_detection)
+    {
+        world.updateFieldGeometry(*field_detection);
+    }
+    world.mutableEnemyTeam()    = new_enemy_team;
+    world.mutableFriendlyTeam() = new_friendly_team;
+    if (new_ball)
+    {
+        world.updateBallState(new_ball->currentState());
+    }
```

Comment:
could we refactor this function into:
```
parseDetections(vision_detection)
applyFilters()
updateGoalies()
updateCurrentWorld()
```
? Seems like it's already basically split up that way.

Reply:
I would generalize `updateGOalies()` to something like `updateRefox` or similar (assuming that's where the data is coming from here).

Reply:
Actually name it whatever is more relevant for the code _as it is now_ and we can change later once we do more gamecontroller integration.

Reply:
refactored it